core/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java

```diff
@@ -348,7 +348,7 @@ public void onFailure(Throwable e) {
                 });
                 success = true;
             } catch (InterruptedException e) {
-                Thread.interrupted();
+                Thread.currentThread().interrupt();
```

Comment:
This puts the burden on the client (or anybody that implements such a listener), which needs to know that we do not restore the interrupt flag and do it in the listener implementation instead. As a consequence of this decision the client will not be able to cancel a bulk request *unless* it is aware of the idiosyncrasies of our client API.

The problem is that `InterruptedException` is not like any other exception and we're really interfering with the expectations of the client here. To me we're assuming too much about the client when we do not restore the flag here.